src/gateway.h

```diff
@@ -54,15 +57,20 @@ void gateway__leader_close(struct gateway *g, int reason);
 
 /**
  * Asynchronous request to handle a client command.
+ *
+ * We also use the handle as a place to save request-scoped data that we need
+ * to access from a callback.
  */
 typedef void (*handle_cb)(struct handle *req, int status, int type);
 struct handle
 {
-	void *data; /* User data */
-	int type;   /* Request type */
+	void *data;                   /* User data */
+	int type;                     /* Request type */
 	struct gateway *gateway;
 	struct buffer *buffer;
 	struct cursor cursor;
+	uint64_t db_id;               /* For use by prepare callback */
+	struct stmt *registered_stmt; /* For use by prepare callback */
```

Comment:
I would probably change this to `size_t stmt_id`, since once a statement gets added to the registry it is normally accessed by ID, rather than by direct memory access. This has the advantage of ensuring consistency, in that you know for sure that the object is actually registered. It is also more consistent with the `db_id` field.